### PR TITLE
fix: handling untagged images when tagging and removing (#1511 + #1512)

### PR DESCRIPTION
**Pull request by iwilltry42**
Ref #1511 & #1512

### Checklist
- [x] The title of this PR would make a good line in Acorn's Release Note's Changelog
- [x] The title of this PR ends with a link to the main issue being address in parentheses, like: `This is a title (#1216)`. [Here's an example](https://github.com/acorn-io/acorn/pull/1199)
- [x] All relevant issues are referenced in the PR description. *NOTE: don't use [GitHub keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) that auto-close issues*
- [x] Commits follow [contributing guidance](https://github.com/acorn-io/acorn/blob/main/CONTRIBUTING.md#commits)
- [x] Automated tests added to cover the changes. If tests couldn't be added, an explanation is provided in the Verification and Testing section
- [ ] Changes to user-facing functionality, API, CLI, and upgrade impacts are clearly called out in PR description
- [ ] PR has at least two approvals before merging (or a reasonable exception, like it's just a docs change)

### CODE REVIEW

**Comment by iwilltry42** [comment]
ImageGet can use some simplification.
I can include this here or in a follow-up.

File: pkg/cli/images_rm.go [hunk 1]
```diff
@@ -34,18 +31,7 @@ func (a *ImageDelete) Run(cmd *cobra.Command, args []string) error {
...
-			opts = append(opts, name.WithDefaultTag(""))
-		}
-
-		// normalize image name (adding :latest if no tag is specified and it's not a digest or potential ID)
-		ref, err := name.ParseReference(image, opts...)
-		if err != nil {
-			return err
-		}
-		deleted, err := c.ImageDelete(cmd.Context(), strings.TrimSuffix(ref.Name(), ":"), &client.ImageDeleteOptions{Force: a.Force})
+		deleted, err := c.ImageDelete(cmd.Context(), args[0], &client.ImageDeleteOptions{Force: a.Force})
```
[the remark this answers is not in the document]
Comment (iwilltry42) [reply pkg/cli/images_rm.go:34]: Whoopsie :grin:

**Comment by g-linville** [comment]
`findImageMatch` was pretty tough for me to understand. Is one of the possible matches it's looking for the tag only? For example, if my input were simply the string `dev` to that function, would it return an image for example `some-registry/some-repo-name:dev`, as long as that was the only image with the tag `dev`? That's what it looks to me like it would do, and imo that would be odd behavior. But maybe I misread it.

**Comment by iwilltry42** [comment]
@g-linville no, tags in Acorn images are the items in the tags field of the image CRD.
They are not just the tags, gut full identifier (registry/repo[:Tag]).

`dev` should be treated as `dev:latest`.

I can add some more tests though to clarify the various cases.

**Comment by iwilltry42** [comment]
@g-linville added some test cases including one for tag-only.l 👍

**Review by g-linville — APPROVED** [review]

File: pkg/server/registry/apigroups/acorn/images/strategy.go [hunk 2]
```diff
@@ -264,27 +269,44 @@ func findImageMatch(images apiv1.ImageList, imageName string) (*apiv1.Image, str
...
 
-		for i, tag := range image.Tags {
-			if tag == imageName {
-				return &image, image.Tags[i], nil
+		// >>> match by tag name
+		for _, tag := range image.Tags {
+			if tag == search {
+				if !canBeMultiple {
+					return &image, tag, nil
+				} else {
```
Comment (thedadams) [inline pkg/server/registry/apigroups/acorn/images/strategy.go:277]: nit: `else` not needed here.

File: pkg/server/registry/apigroups/acorn/images/strategy.go [hunk 3]
```diff
@@ -264,27 +269,44 @@ func findImageMatch(images apiv1.ImageList, imageName string) (*apiv1.Image, str
...
-				imageParsedTag, err := name.NewTag(tag, name.WithDefaultRegistry(""))
+				imageParsedTag, err := name.NewTag(tag, name.WithDefaultRegistry(""), name.WithDefaultTag("")) // no default here, as we also have repo-only tag items
 				if err != nil {
 					continue
 				}
 				if imageParsedTag.Name() == tagName {
-					return &image, tag, nil
+					if !canBeMultiple {
+						return &image, tag, nil
+					} else {
```
Comment (thedadams) [inline pkg/server/registry/apigroups/acorn/images/strategy.go:292]: nit: `else` not needed here.

**Review by thedadams — APPROVED** [review]

File: pkg/server/registry/apigroups/acorn/images/strategy.go (hunk 2 above)
Comment (iwilltry42) [reply pkg/server/registry/apigroups/acorn/images/strategy.go:277]: true :)

**Pull request merged**